### PR TITLE
Rearrange expand()

### LINKED ISSUE #222: Tweak the expand() function again to reduce copying
I've discovered that the copying involved in the 4-pass algorithm may be a non-negligible part of the total decrypt time.

We can eliminate some of the copying in 4 pass by putting the length and index fields at the end of the output of expand, e.g.

OLD
expand(0x484848, 7, 1) = (0x07, 0x01, 0x48, 0x48, 0x48, ....)
NEW
expand(0x484848, 7, 1) = (0x48, 0x48, 0x48, .... 0x07, 0x01)

in the likely case where server_id_len < nonce_len, this means that the decoder can just pass server_id_len bytes of left_1 to the caller, rather then copying it into a new buffer.

### PR DESCRIPTION
**Pull request by martinduke**
by moving the arbitrary location of some bytes around, we can eliminate a copy and improve performance on 3-pass decryption by about 7%.

I would very much appreciate someone verifying my test vectors.

Fixes #222

### CODE REVIEW

File: draft-ietf-quic-load-balancers.md [hunk 1]
```diff
@@ -497,21 +497,20 @@ The expand(length, pass, input_bytes) function concatenates three arguments and
...
 The second argument is an 8-bit integer that is the 'pass' of the algorithm, and
-forms the second-most significant octet of the output.
+forms the sixteenth (least significant) octet of the output.
 
 The third argument is a variable-length stream of octets, which is copied into
-the third-most significant octet of the output and beyond. The length of this
-octet stream is half the 'length', rounded up. All remaining octets of the
-output are zero.
+the most significant octets of the output. The length of this octet stream is
+half the 'length', rounded up. All remaining octets of the output are zero.
```
Comment (huitema) [inline draft-ietf-quic-load-balancers.md:509]: This appears to be half an edit. The syntax as you express it appears to be:
~~~
ExpandResult {
      Bytes(...),
      ZeroPad(...),
      length(8),
      pass(8)
}
~~~

In which:

* 'Bytes' are the N most significant octets of the output. The number N is half the 'length', rounded up, thus a number between 1 and 14.
* 'Zeropad' is a set of 14-N zeroes,
* 'length' is an 8-bit integer that reports the sum of the configured nonce length and server id length in octets, and forms the fifteenth octet of the output. The 'length' argument MUST NOT exceed 28.
* 'pass' is an 8-bit integer that reports the 'pass' argument of the algorithm, and
forms the sixteenth (least significant) octet of the output.

**Review by huitema — COMMENTED** [review]
Yes, I see how this new syntax leads to simpler code. But the text could be improved, see suggestion.

File: draft-ietf-quic-load-balancers.md (hunk 1 above)
Comment (martinduke) [reply draft-ietf-quic-load-balancers.md:525]: OK, I took this working, minus some corrections. Length has to be between 5 and 19, meaning N is between 3 and 10.

File: draft-ietf-quic-load-balancers.md [hunk 2]
```diff
@@ -496,22 +496,37 @@ Two functions are useful to define:
...
 outputs 16 zero-padded octets.
 
-The first argument 'length' is an 8-bit integer that reports the sum of the
-configured nonce length and server id length in octets, and forms the most
-significant octet of the output. The 'length' argument MUST NOT exceed 28.
+The output of expand is as follows:
 
-The second argument is an 8-bit integer that is the 'pass' of the algorithm, and
-forms the second-most significant octet of the output.
+~~~psuedocode
```
Comment (huitema) [inline draft-ietf-quic-load-balancers.md:501]: Typo.

File: draft-ietf-quic-load-balancers.md [hunk 3]
```diff
@@ -496,22 +496,37 @@ Two functions are useful to define:
...
+* 'input_bytes' is drawn from one half of the plaintext. It forms the N most
+significant octets of the output, where N is half the 'length' argument, rounded
+up, and thus a number between 3 and 10, inclusive.
 
-The third argument is a variable-length stream of octets, which is copied into
-the third-most significant octet of the output and beyond. The length of this
-octet stream is half the 'length', rounded up. All remaining octets of the
-output are zero.
+* 'Zeropad' is a set of 14-N zeroes.
+
```
Comment (huitema) [inline draft-ietf-quic-load-balancers.md:517]: Reading again, I would write "a set of 14-N octets set to zero."

**Review by huitema — APPROVED** [review]
Looks good now.

**Pull request merged**